src/esp32/pwm.rs

```diff
@@ -0,0 +1,209 @@
+        1 => LedcTimerDriver::new(unsafe { TIMER1::new() }, &pwm_tconf)?,
+        2 => LedcTimerDriver::new(unsafe { TIMER2::new() }, &pwm_tconf)?,
+        3 => LedcTimerDriver::new(unsafe { TIMER3::new() }, &pwm_tconf)?,
+        _ => anyhow::bail!("Could not find TIMER{:}", timer_number),
+    };
+    Ok(match channel {
+        PwmChannel::C0 => LedcDriver::new(unsafe { CHANNEL0::new() }, timer, pin)?,
+        PwmChannel::C1 => LedcDriver::new(unsafe { CHANNEL1::new() }, timer, pin)?,
+        PwmChannel::C2 => LedcDriver::new(unsafe { CHANNEL2::new() }, timer, pin)?,
+        PwmChannel::C3 => LedcDriver::new(unsafe { CHANNEL3::new() }, timer, pin)?,
+        PwmChannel::C4 => LedcDriver::new(unsafe { CHANNEL4::new() }, timer, pin)?,
+        PwmChannel::C5 => LedcDriver::new(unsafe { CHANNEL5::new() }, timer, pin)?,
+        PwmChannel::C6 => LedcDriver::new(unsafe { CHANNEL6::new() }, timer, pin)?,
+        PwmChannel::C7 => LedcDriver::new(unsafe { CHANNEL7::new() }, timer, pin)?,
+    })
+}
+
+pub struct PwmDriver<'a> {
+    // the timer property on this LedcDriver is unreliable due to the logic below
+    // in set_frequency
+    ledc_driver: LedcDriver<'a>,
+    timer_number: usize,
+    channel: PwmChannel,
+}
+impl<'a> PwmDriver<'a> {
+    fn new(pin: AnyIOPin, channel: PwmChannel, starting_frequency_hz: u32) -> anyhow::Result<PwmDriver<'a>> {
+        let mut ledc_manager = LEDC_MANAGER.lock().unwrap();
+        let timer_number = ledc_manager.take_timer(starting_frequency_hz, channel)?;
+        let ledc_driver = get_ledc_driver_by_channel(channel, timer_number, starting_frequency_hz, pin)?;
+
+        Ok(PwmDriver {
+            ledc_driver,
+            timer_number,
+            channel
+        })
+    }
+
+    pub fn set_duty_pct(&mut self, pct: f64) -> anyhow::Result<()> {
+        let max_duty = self.ledc_driver.get_max_duty();
+        self.ledc_driver
+            .set_duty(((max_duty as f64) * pct.abs()).floor() as u32)?;
+        Ok(())
+    }
+
+    pub fn get_duty_pct(&self) -> f64 {
+        let max_duty = self.ledc_driver.get_max_duty();
+        (self.ledc_driver.get_duty() as f64) / (max_duty as f64)
+    }
+
+    pub fn get_frequency(&self) -> u32 {
+        unsafe { ledc_get_freq(SpeedMode::LowSpeed.into(), self.ledc_driver.timer()) }
+    }
+
+    pub fn set_frequency(&mut self, frequency_hz: u32) -> anyhow::Result<()> {
+        let mut ledc_manager = LEDC_MANAGER.lock().unwrap();
+        ledc_manager.drop_timer(self.timer_number, self.channel);
+        let timer_number = ledc_manager.take_timer(frequency_hz, self.channel)?;
+        let timer: ledc_timer_t = (timer_number as u8).into();
+        match esp_idf_sys::esp!(unsafe {
+            ledc_set_freq(SpeedMode::LowSpeed.into(), timer, frequency_hz)
+        }) {
+            Ok(val) => { Ok(val) },
+            Err(err) => {
+                ledc_manager.drop_timer(timer_number, self.channel);
```

Comment:
why is it potentially called twice?

Reply:
because according to the LedcManager, the timer has been taken in L. 100, but I can just move L. 99 down